Write & mmap chunks in the background

prymitive opened the pull request:
Mmapping chunks is expensive and so we should avoid having this done on the hot path to avoid blocking scrapes and rule evaluations.
This change allows tsdb to keep most recent chunks in-memory and write & mmap them on a schedule by a background goroutine.

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

This is an attempt to address #10377, early testing shows that it does in fact resolve most of the issue for me.

@codesome / @replay

prymitive commented:
![Screenshot from 2022-05-17 17-26-00](https://user-images.githubusercontent.com/114288/168861469-f175350e-30d5-49b0-be6d-f8aebcb31753.png)
![Screenshot from 2022-05-17 17-25-45](https://user-images.githubusercontent.com/114288/168861472-370c672d-eeec-41da-a654-4b393332ed8f.png)
![Screenshot from 2022-05-17 17-25-09](https://user-images.githubusercontent.com/114288/168861474-dab29cb4-b465-4423-904c-239abcc2aeb9.png)
![Screenshot from 2022-05-17 17-24-42](https://user-images.githubusercontent.com/114288/168861478-288f834a-38f3-4b65-b440-658d0007dcfe.png)

roidelapluie commented:
/prombench main

prymitive commented:
FYI the main improvement here, I think, is the way writes are scheduled.
There's now one big loop that iterates all chunks, locks each one and mmaps them. So we have strict ordering which helps to avoid getting blocked and waiting long for a lock.
Before chunk appends would trigger mmap, so at some point we would have huge number of chunks competing for a lock, hence some of them would get unlucky and stuck waiting on a lock, blocking reads and appends to such chunk.

prymitive commented:
More screenshots with Prometheus metrics after deploying latest patch.

Query and rule evaluation duration did improve for p90:
![Screenshot from 2022-05-17 21-14-26](https://user-images.githubusercontent.com/114288/168902071-82436f19-b53e-4dd4-8ed5-008a59164110.png)
![Screenshot from 2022-05-17 21-14-08](https://user-images.githubusercontent.com/114288/168902074-d0f3caa6-d2b2-454d-b391-011a181ace1d.png)

And I'm no longer seeing missed evaluations:
![Screenshot from 2022-05-17 21-13-47](https://user-images.githubusercontent.com/114288/168902075-b39a65d4-0842-4121-b640-898e35bb4cbb.png)

Sample rate is more stable, before we had a dip when scrapes get blocked followed by a spike as it got unblocked:
![Screenshot from 2022-05-17 21-13-06](https://user-images.githubusercontent.com/114288/168902077-9f3410c9-829d-4952-a24a-c06f93e313fd.png)

And rule evaluation duration is now more stable, no more big spikes in duration as rules get blocked:
![Screenshot from 2022-05-17 21-12-45](https://user-images.githubusercontent.com/114288/168902078-d8b5b90c-767b-4282-ac9d-15fc288ce1a2.png)

But there is a slight memory increase, which I need to look into and see if there's anything I can do about it:
![Screenshot from 2022-05-17 21-12-15](https://user-images.githubusercontent.com/114288/168902082-4d5e690c-eda0-4c7e-b291-b91c51a57ab2.png)

roidelapluie commented:
Note that the rule until now is that TSDB only uses one core, therefore one go routine.

prymitive commented:
> Note that the rule until now is that TSDB only uses one core, therefore one go routine.

chunk write queue, which was added a few releases ago and disabled recently as it caused us to notice big performance degradation in the linked issue, already runs an extra goroutine to process writes

brian-brazil commented:
Each background thread is 1.33 cores that every user needs to provision, as you need to presume the worst case that they're all active at the same time.

prymitive commented:
> Each background thread is 1.33 cores that every user needs to provision, as you need to presume the worst case that they're all active at the same time.

As I've mentioned on https://github.com/prometheus/prometheus/issues/10377#issuecomment-1129071559 running mmap off a goroutine scheduled every N minutes is arbitrary. The goal was to move this whole process away from append() path so it doesn't block scrapes & rules. I wasn't sure what's the best moment to mmap extra chunks and so I went with a background goroutine. Happy to remove this goroutine but I'm not sure where & when mmap should be triggered (assuming the rest of the patch can stay).

brian-brazil commented:
Could we do it at the previous head compaction? After the very first, we should have a good idea of how many we'll need.

prymitive commented:
Do you mean here - https://github.com/prometheus/prometheus/blob/707600d84f5567354ef48d4593ddab84c4837811/tsdb/db.go#L887 ?

prymitive commented:
I wonder if mmap should also be forced in `ChunkSnapshot()` - https://github.com/prometheus/prometheus/blob/9558b9b54bd3d0cb1d63b9084f8cbcda6b0d72fb/tsdb/head_wal.go#L581

prymitive commented:
Deployed latest patch to another instance and it also show improvement. No longer seeing rule evaluation duration spikes leading to missed iterations.
Memory usage looks similar to vanilla v2.35.0

![Screenshot from 2022-05-18 14-54-01](https://user-images.githubusercontent.com/114288/169058138-bd1e19b3-0b91-4b95-bb5b-953b9e74adcb.png)
![Screenshot from 2022-05-18 14-53-38](https://user-images.githubusercontent.com/114288/169058150-9047acfb-3955-4413-88b0-5b325f6a9226.png)
![Screenshot from 2022-05-18 14-53-19](https://user-images.githubusercontent.com/114288/169058159-72997f1a-0015-4237-87a6-fe73e9ecf9b3.png)
![Screenshot from 2022-05-18 14-52-59](https://user-images.githubusercontent.com/114288/169058163-9ef8f1ae-42c6-43b4-be75-18be7974469b.png)
![Screenshot from 2022-05-18 14-52-21](https://user-images.githubusercontent.com/114288/169058169-6a6965c6-19c7-4cbc-9f97-c56c8d297b25.png)

roidelapluie commented:
/prombench cancel

codesome reviewed (commented):
We already have a async queue that does this (See https://github.com/prometheus/prometheus/pull/10051). It works but had some bottleneck somewhere on locks.

I have not read your PR in detail yet, but I see that you have a loop to go over all series and m-map them. I don't think we want to m-map all series at the same time. Also, I would avoid making the `headChunk` a slice, it is simpler to work with a single chunk here.

If it interests you, would you like to take a look at the existing queue and see how we can improve it? (We disabled it by default here https://github.com/prometheus/prometheus/pull/10425)

(Edit: I see that you reported the bottlenecks in https://github.com/prometheus/prometheus/issues/10377)

codesome commented:
I see that you have already done some investigation in https://github.com/prometheus/prometheus/issues/10377. Did you have the queue enabled? Did this patch with the queue make the situation better?

prymitive commented:
> I see that you have already done some investigation in #10377. Did you have the queue enabled? Did this patch with the queue make the situation better?

Enabling chunk_write_queue causes severe degradation, which is why #10377 was raised. The patch for queue locks didn't show any noticeable performance improvement for my workload.
Without queue enabled servers I run that have 20+ million time series will show a minute or two minute long performance issue, as new chunks are being mmapped. Enabling the queue turns it into 5+ minute window of poor performance.

I did dig a bit more into both the code and history of changes around mmapping chunks, and I did realize that the queue added in #10051 was aimed at fixing the exact issue I'm observing, but it seems that it was focused at fixing remote write problems, so it might have improved writes but with unfortunate side effect of degrading reads under some conditions.

I see this patch as an alternative solution to the same problem. In both cases you keep chunks in memory until you're ready to mmap them. The difference is that the queue variant moves in-memory chunks to a new single component with additional locking that might have new/different bottlenecks, while I was trying to keep everything as parallel and distributed as it is (since that works well) and do the mmapping in a more ordered and controllable way.
With the queue I hit an hour on the wall clock and all in-memory are suddenly sent to the queue where they fight for locks. With this patch there's a loop that writes one chunk at a time, minimising any potential for lock congestion to minimum.

prymitive commented:
> I see that you have already done some investigation in #10377. Did you have the queue enabled? Did this patch with the queue make the situation better?

I *think* that the queue doesn't really help here because it calls `WriteChunk()` which then needs to hold a lock

https://github.com/prometheus/prometheus/blob/59727ab0eb48802690e47e2920934e1b342894cf/tsdb/chunks/head_chunks.go#L382-L383

 which essentially serialised all writes. Without the queue this path is more expensive since we need to do the actual write, but it seems that this code path is expensive enough even if we just queue it, rather than do a full write.
 
 When tsdb block needs to be written and suddenly we need to create a new chunk for every time series being scraped and rule being evaluated we hit a race window. Let's say scrapes and rules have 1 minute interval. We hit tsdb block max duration and start creating new chunks for everything. We have a rule evaluation that produces a sample or a scrape that returns a sample - we need to append it.
As we call append() tsdb cuts a new chunk.
Since cutting a new chunks is blocking (because there are locks like the one mentioned above) we are suddenly against the clock - our append is potentially blocked by appends on other chunks.
And so if we get unblocked within a minute then everything works.
But if our append gets blocked for more than a minute then we've missed scrape/rule interval.
And that's just for the write path. There are also reads to consider.
Our chunk might need to be read for queries or rule evaluations, if it's blocked for more than 2 minutes then any query using it will timeout.
So it's very important to keep any blocking under scrape/rule interval and query timeout. Otherwise we can't keep up.

The problem I'm hitting is that I have an instance with 20-26M time series. Since WriteChunk is essentially serialised then the more chunks I have in Prometheus the longer any blocking will last. This is made worse the more rules using chunks produced by other rules I have. So eventually I will hit that 1 minute (rule interval) and 2 minute (query timeout) marks and see failures.

Since the problem is with appends getting blocked, which only happens when we need to cut new chunks, and the queue doesn't seem enough to remediate this issue, I've decided that the easiest way to deal with it is to get rid of blocking on append() path. If we don't need to block and wait for mmap when appending then always have same performance level.

Now since mmapping of chunks was added to lower memory usage we still do want to mmap it all, but this patch moves that to be run around GC time as suggested by @brian-brazil.
Mmapping is done in a loop, one chunk at a time, this has the advantage that we minimise any blocking, since we mmap one chunk at a time the lock in WriteChunk is unlikely to be held, and even if it is that's gonna be a lot fewer chunks locking it then during that minute when tsdb block is written and all time series need new chunks.
Now it was also mentioned that we want to avoid having more than a single goroutine for tsdb, which this patch does have, but the queue code needs to spin additional goroutine to service writes, which seem to violate this requirement.

codesome reviewed (commented):
The approach is interesting. I would like to see if 2.37.0 makes this better with a bigger mmap queue size (see [this comment](https://github.com/prometheus/prometheus/issues/10377#issuecomment-1218053083)) before going ahead with this

codesome commented:
This approach is interesting and something we can consider. I would like us to wait until https://github.com/prometheus/prometheus/pull/11075 is merged because doing it the other way round will be more challenging.

bwplotka commented:
SGTM, great job on this!

prymitive commented:
Rebased on top of v2.39.0

codesome commented:
Hey @prymitive, just popping in here to say that I haven't forgotten about this. I hope to get back to this soon. Multiple conferences (and hence conference driven development) taking up my time at the moment :)

prymitive commented:
> Hey @prymitive, just popping in here to say that I haven't forgotten about this. I hope to get back to this soon. Multiple conferences (and hence conference driven development) taking up my time at the moment :)

Thanks.
I did rebase this on top of 2.39, which was a big change and required a bit of a rewrite in some places.
I started deploying 2.39.1+this patch to my servers. So far no issues identified.

prymitive commented:
Any chance this be moved forward now?

codesome commented on tsdb/db.go:
```diff
@@ -1238,6 +1246,14 @@ func (db *DB) compactHead(head *RangeHead) error {
 	return nil
 }
 
+func (db *DB) mmapHead() {
```
This is a Head specific thing. We should avoid spilling that into db.go. And I don't think we should call it from here anywhere; Head should take care of calling it regularly.

prymitive replied:
Moved to HEAD

codesome commented on tsdb/db_test.go:
```diff
@@ -1493,6 +1493,7 @@ func TestSizeRetention(t *testing.T) {
 		}
 	}
 	require.NoError(t, headApp.Commit())
+	db.Head().series.mmapHeadChunks(db.Head().chunkDiskMapper)
```
It would be simpler if we just added a mmapHeadChunks method to the Head, and that does all this heavy lifting of knowing what to call and how.

prymitive replied:
Moved it all into Head.mmapHeadChunks()

codesome reviewed (commented):
I was reviewing the PR, but I stopped after noticing something: are we m-mapping only right before the compaction or am I missing someplace where the m-mapping is triggered as well?

prymitive commented:
> I was reviewing the PR, but I stopped after noticing something: are we m-mapping only right before the compaction or am I missing someplace where the m-mapping is triggered as well?

Initially I've added a goroutine that mmaps on a schedule. But there were comments that TSDB is not allowed to spin additional goroutines - https://github.com/prometheus/prometheus/pull/10709#issuecomment-1129715073 - and so I removed that and instead hooked mmapping around compaction.

But frankly I don't know what is the best moment to mmap chunks.
For majority of chunks it's probably best to mmap after block range, so every 2h, right after all memSeries get a new head chunk due to reaching max allowed time.
But depending on the setup there might be a lot of jobs with short scrape interval that's not very well aligned with maxt and so they might produce more than one chunk during a single block range, and for these it would make sense to try and mmap chunks more often.

The current code effectively runs mmap every ~1 minute on my servers, with mmaping taking 0-15s depending on the number of time series and usually mmaps 300-1500 chunks.
This doesn't cause any performance issues that I observed but it's probably more often than it needs to be.

Not sure where and when it should be ideally triggered.

prymitive commented on tsdb/head.go:
```diff
@@ -1906,11 +1921,17 @@ func (s *memSeries) maxTime() int64 {
+	}
+	if removedInOrder > 0 {
+		s.firstChunkID += chunks.HeadChunkID(removedInOrder + len(s.mmappedChunks))
+		s.headChunks = append(s.headChunks[:0], s.headChunks[removedInOrder:]...)
```
Refactored

codesome commented:
Oh, I see it now. We will trigger the function everytime Compact() is called, and does not depend on compaction happening.

I would suggest moving that call inside `DB.run()`, and inside the reloadBlocks() case. It would be good to not depend on the compaction call cycle and reliably trigger it every minute. If the frequency turns out to be too much, we can always tweak it in the future.

I will review rest of the code soon.

codesome commented on tsdb/head_read.go:
```diff
@@ -319,27 +327,25 @@ func (s *memSeries) chunk(id chunks.HeadChunkID, chunkDiskMapper *chunks.ChunkDi
 	// is len(s.mmappedChunks), it represents the next chunk, which is the head chunk.
 	ix := int(id) - int(s.firstChunkID)
-	if ix < 0 || ix > len(s.mmappedChunks) {
+	if ix < 0 || ix > len(s.mmappedChunks)+len(s.headChunks) {
```
`ix` is 0 based. Previously `len(s.mmappedChunks)` meant the last head chunk. Now `len(s.mmappedChunks)+len(s.headChunks)-1` is the last chunk.
```suggestion
	if ix < 0 || ix > len(s.mmappedChunks)+len(s.headChunks)-1 {
```

Because tests didn't fail, we need to add a test case such that `return s.headChunks[ix], false, nil` would have panicked.

codesome commented on tsdb/head_read.go:
```diff
@@ -646,8 +652,13 @@ func (s *memSeries) iterator(id chunks.HeadChunkID, isoState *isolationState, ch
+		ix -= len(s.mmappedChunks)
+
+		for j, d := range s.headChunks {
+			totalSamples += d.chunk.NumSamples()
+			if j < ix {
+				previousSamples += d.chunk.NumSamples()
+			}
```
We can call `d.chunk.NumSamples()` only once by assigning it to a variable. It actually reads the chunk bytes and decodes the number.

codesome commented on tsdb/head_test.go:
```diff
@@ -727,49 +727,111 @@ func TestHead_Truncate(t *testing.T) {
-	}()
-	const chunkRange = 2000
+	type testCaseT struct {
+		chunkRange     int
```
`chunkRange` is the same for all cases. So we can remove this field from here.

codesome commented on tsdb/head_test.go:
```diff
@@ -727,49 +727,111 @@ func TestHead_Truncate(t *testing.T) {
-	require.NoError(t, err)
+			// append samples that will end up in mmapped chunks
+			if tc.mmappedSamples > 0 {
+				for i := 0; i < tc.mmappedSamples; i += 5 {
+					ok, _ := s.append(int64(i), float64(i), 0, chunkDiskMapper, int64(tc.chunkRange))
+					require.True(t, ok, "sample append failed")
+				}
+				s.mmapHeadChunks(chunkDiskMapper)
+				require.Greater(t, len(s.mmappedChunks), 0)
+			}
 
-	s.truncateChunksBefore(2000, 0)
+			// append samples that will stay in the head
+			if tc.headSamples > 0 {
+				for i := tc.mmappedSamples; i < tc.mmappedSamples+tc.headSamples; i += 5 {
+					ok, _ := s.append(int64(i), float64(i), 0, chunkDiskMapper, int64(tc.chunkRange))
+					require.True(t, ok, "sample append failed")
+				}
+				require.Greater(t, len(s.headChunks), 0)
+			}
```
You can simplify this by having a loop from 0 to tc.headSamples with a condition inside like
```
if i == tc.mmappedSamples {
    // mmap
    // check for >0
} 
```

(while you do that, can you change the comment to start with a capital letter where possible and end with a `.`?)

codesome commented on tsdb/head_test.go:
```diff
@@ -727,49 +727,111 @@ func TestHead_Truncate(t *testing.T) {
+			require.NoError(t, err)
+			require.Equal(t, lastChunk, chk)
+
+			// verify that we can read
```
```suggestion
			// Verify that we can read.
```

codesome reviewed (commented):
Nice work! I just have a bunch of small comments now.

codesome commented:
/prombench main

One last benchmark after some recent changes.

codesome commented:
/prombench restart main

codesome commented:
/prombench cancel

codesome commented:
/prombench main

Some CI issue earlier and I messed up the CI by restarting it. Trying again.

codesome commented:
Prombench start CI failed again with the same error:
```
09:21:36 gke.go:594: error while applying a resource err:error applying 'manifests/prombench/benchmark/6_loadgen.yaml' err:Request for 'applying deployment:prometheus-loadgen-querier' hasn't completed after retrying 50 times
```

But we still have a load of around 1m series. So will let it run for a while.

codesome commented on tsdb/head.go:
```diff
@@ -1550,6 +1554,25 @@ func (h *Head) getOrCreateWithID(id chunks.HeadSeriesRef, hash uint64, lset labe
+		}
+		h.series.locks[i].RUnlock()
+	}
+	if mmapped > 0 {
+		level.Info(h.logger).Log("msg", "Finished mmapping head chunks", "chunks", mmapped, "duration", time.Since(start).String())
+	}
```
Alternatively, we can make it a debug log

prymitive replied:
I was thinking of turning this into a metric instead

codesome commented:
There is a sustained 5% increase in the memory RSS, I believe it is the switch to a slice of headChunks and other factors. CPU usage is also slightly up.

I would like to borrow @bboreham's expertise here. @bboreham: This basically helps with latency spikes as described in https://github.com/prometheus/prometheus/issues/10377. Apparently, the async mmap queue was not helpful. This PR triggers the mmap of in-memory chunks in the background by holding a slice of in-memory chunks per series. Does the memory/CPU increase look concerning or is there a way to optimize it?

codesome commented:
/prombench cancel

I think we have enough data for now.

prymitive commented:
> There is a sustained 5% increase in the memory RSS, I believe it is the switch to a slice of headChunks and other factors. 

Slice needs extra 24 bytes if I recall. That's 24MB per 1 million time series.
Last time I checked (https://github.com/prometheus/prometheus/pull/11003#issuecomment-1181676566) memSeries was using 224 bytes, so extra 24 bytes is 10% more just for baseline memory usage.
That being said most memory in memSeries will be eaten by samples and labels. In my setup I usually see around 3.2KB per memSeries and 24 bytes is <1% memory usage bump when you compare it with that number. Real world overhead depends on how big your memSeries are really.
But the bottom is obvious here - having a slice here uses more memory.

But at the same time there was another slice added for out of order chunks, and we also have another slice for mmapped chunks. An obvious saving would be if we managed to keep all different chunks on a single slice, regardless if they are head, mmapped or OoO chunks. That would require a lot more changes but there are bits of code that I found difficult to get my head around quickly as they need to get the right chunk from passed offset (see `tsdb/head_read.go` changes) and having a single slice could potentially simplify some of that, which would be great.

Chunk write queue and other alternative solutions are likely to eat up some extra memory too, but those might be a bit harder to calculate exact memory usage growth due to more dynamic nature of it. Plus with queues you don't pay that extra memory usage all the time (except for the base structure to have the queue running), so from purely "bytes payed" PoV slice is more of a small but permanent cost that scales with the number of time series.
The upside is that it removes blocking behaviour from the hot path. Another solution could be to keep two pointers on memSeries: one for current HEAD chunk and one for previous one that should be mmapped. That only adds 8 bytes. The reason I went for the slice is that it felt safer - with two pointers I need to ensure "previous" chunk is mmapped before current HEAD one gets full, otherwise we need to block again - there's no such worry about slices.

bboreham commented:
Since you changed a single pointer to a slice, it should be +16 bytes per series. (When looking at rss this number is doubled by Go's garbage-collector).

Would it work as a linked list?  I have pondered making that change for `mmappedChunks` and `oooMmappedChunks`.

prymitive commented:
> Would it work as a linked list?

I cannot think of any reason why it wouldn't. Reading from memSeries would become a bit more complicated, since referencing chunks via `series.headChunks[3]` wouldn't be possible and we would have to iterate over the list until we get to the 4th chunk. But that's the main complication I can think of. No idea how that would impact read performance.

bboreham commented:
I think in practice those indices are only ever used from an iterator stepping through the set of chunks, but the code assumes this can be effectively expressed by an integer.

prymitive commented:
Here's a linked list version of this PR - #11818

prymitive commented:
Closing this one, there's #11818 instead
